inference-engine/thirdparty/clDNN/src/network.cpp

```diff
@@ -393,23 +396,131 @@ void network_impl::set_input_data(const primitive_id& id, memory::ptr data) {
+            if (eng.is_the_same_buffer(mem_orig, usr_prim->output_memory())) {
+                chain.push_back(usr_prim);
+            }
+        }
+    };
+
+    auto mdata_ptr = std::dynamic_pointer_cast<mutable_data_inst>(p_inst);
+    if (p_inst->can_be_optimized()) {
+        candidates.push(p_inst);
+    } else if (mdata_ptr != nullptr) {
+        add_mdata_chain(p_inst);
+    }
+
+    // find all dependencies that are 'optimized'
+    while (!candidates.empty()) {
+        auto& cand = candidates.top();
+        candidates.pop();
+        const auto& mem_cand = cand->output_memory();
+        if (eng.is_the_same_buffer(mem_orig, mem_cand)) {
+            auto nc_cand = std::const_pointer_cast<primitive_inst>(cand);
+            chain.push_back(nc_cand);
+            auto mdata_ptr = std::dynamic_pointer_cast<mutable_data_inst>(nc_cand);
+            if (mdata_ptr != nullptr) {
+                add_mdata_chain(nc_cand);
+            }
+        }
+
+        for (auto& dep : cand->dependencies()) {
+            if (dep->can_be_optimized()) {
+                candidates.push(dep);
+            } else {
+                const auto& mem_dep = dep->output_memory();
+                if (eng.is_the_same_buffer(mem_orig, mem_dep)) {
+                    auto nc_dep = std::const_pointer_cast<primitive_inst>(dep);
+                    chain.push_back(nc_dep);
+                    auto mdata_ptr = std::dynamic_pointer_cast<mutable_data_inst>(nc_dep);
+                    if (mdata_ptr != nullptr) {
+                        add_mdata_chain(nc_dep);
+                    }
+                }
+            }
+        }
+    }
+
+    if (chain.empty()) {
+        return _output_chains.end();
+    } else {
+        std::sort(chain.begin(), chain.end());
+        chain.erase(std::unique(chain.begin(), chain.end()), chain.end());
+        return _output_chains.insert({ p_inst->id(), chain }).first;
+    }
+}
+
+void network_impl::set_output_memory(const primitive_id& id, memory::ptr mem_new) {
+    std::shared_ptr<primitive_inst> p_inst;
+
+    p_inst = find_primitive(id);
+
+    if (p_inst == nullptr)
         throw std::runtime_error("topology doesn't contain primitive: " + id);
 
-    auto iter = std::find(_outputs.begin(), _outputs.end(), primitive_inst);
+    auto iter = std::find(_outputs.begin(), _outputs.end(), p_inst);
     if (iter == _outputs.end())
         throw std::runtime_error("primitive: " + id + " is not a network output");
 
-    auto output = std::static_pointer_cast<input_layout_inst>(primitive_inst);
-
     // Wait for previous execution completion
     reset_execution(true);
-    output->set_output_memory(mem);
+
+    auto& eng = get_engine();
+    // locate primitive chain for this output
+    // if no chain found - add it
+    auto o_iter = _output_chains.find(id);
+    if (o_iter == _output_chains.end()) {
+        o_iter = add_output_chain(p_inst);
+    }
+
+    if (o_iter == _output_chains.end()) {
```

Comment:
Maybe it's better to make `set_output_memory` virtual and have overload for loop?
I'd want the code below to look like
```cpp
for (auto& prim: o_iter->second) {
    prim->set_output_memory(mem_new);
}
```

Reply:
Fixed. Overloaded **set_output_memory()** for **loop_inst** and **mutable_data_inst**.